core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java

```diff
@@ -909,7 +917,8 @@ public PersistenceState getPersistenceState(long fileId) throws FileDoesNotExist
         inode = inodePath.getInode();
         if (inode.isDirectory()
             && listStatusOptions.getLoadMetadataType() != LoadMetadataType.Always
-            && ((InodeDirectory) inode).isDirectChildrenLoaded()) {
+            && ((InodeDirectory) inode).isDirectChildrenLoaded()
+            && !listStatusOptions.isRecursive()) {
```

Comment:
will a recursive listing always load from the ufs even for `LoadMetadataype.Once`?

Reply:
Maybe this is the same question, but how does `recursive` interact with load metadata?

Reply:
I initially thought `recursive` would force us to load because our only way to detect if a subtree is loaded once is by calling `isDirectChildrenLoaded`, and that does not really tell us about grandchildrens etc. but if there is a way to respect `loadMetadata` flag, specifically `LoadMetadataype.Once`, I think that would be good..

Reply:
sounds like the `loaded` bit needs to be modified so that it can give your more info: `never loaded`, `direct children loaded` or `recursively loaded`

Reply:
@gpang I was talking about inode.mDirectChildrenLoaded which is a single bit. note: "loaded" vs load behavior. makes sense?

Reply:
addressed as discussed.